### PR TITLE
no-jira: Power VS: Refactor `InfraReady` to use metadata client

### PR DESCRIPTION
**Pull request by mjturek**
`InfraReady` recently triggered the gocyclo linter. This patch refactors it.

### CODE REVIEW

File: pkg/infrastructure/powervs/clusterapi/powervs.go [hunk 1]
```diff
@@ -98,118 +95,114 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
...
+	}
+
+	// Step 2: See which ports are already allowed.
+	missingPorts, err := findMissingSecurityGroupRules(ctx, in, *powerVSCluster.Status.VPC.ID)
+	if err != nil {
+		return fmt.Errorf("failed to find missing security group rules: %w", err)
+	}
+
+	// Step 3: Add to security group rules
+	for port := range missingPorts {
```
Comment (r4f4) [inline pkg/infrastructure/powervs/clusterapi/powervs.go:123]:
```suggestion
	for port := range missingPorts {
	    port := port // TODO: remove when using golang 1.22+
```
Loop variable is reused up to golang 1.21 so without a local copy you're always pointing to the same location.

File: pkg/asset/installconfig/powervs/metadata.go [hunk 2]
```diff
@@ -282,3 +285,77 @@ func (m *Metadata) GetDNSServerIP(ctx context.Context, vpcName string) (string,
...
+		Steps:    math.MaxInt32,
+	}
+
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+		err2 := m.client.AddSecurityGroupRule(ctx, vpcID, rule)
+		if err2 == nil {
+			return true, nil
+		}
+		return false, err2
+	})
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:344]: If you're not adding any retry logic:
```suggestion
	err := m.client.AddSecurityGroupRule(ctx, vpcID, rule)
```

File: pkg/asset/installconfig/powervs/metadata.go [hunk 3]
```diff
@@ -282,3 +285,77 @@ func (m *Metadata) GetDNSServerIP(ctx context.Context, vpcName string) (string,
...
+		Steps:    math.MaxInt32,
+	}
+
+	err = wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+		err2 := m.client.CreateDNSRecord(ctx, m.PublishStrategy, instanceCRN, m.BaseDomain, hostname, destHostname)
+		if err2 == nil {
+			return true, nil
+		}
+		return false, err2
+	})
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:309]: If you're not adding any retry logic:
```suggestion
	err = m.client.CreateDNSRecord(ctx, m.PublishStrategy, instanceCRN, m.BaseDomain, hostname, destHostname)
```

Comment (mjturek) [reply pkg/asset/installconfig/powervs/metadata.go]: done

**Comment by r4f4** [comment]
/hold
for https://github.com/openshift/installer/pull/8331

File: pkg/asset/installconfig/powervs/metadata.go [hunk 4]
```diff
@@ -282,3 +285,84 @@ func (m *Metadata) GetDNSServerIP(ctx context.Context, vpcName string) (string,
...
+			Cap:      leftInContext(ctx),
+			Steps:    math.MaxInt32}
+
+	var lastErr error
+	err = wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+		lastErr = m.client.CreateDNSRecord(ctx, m.PublishStrategy, instanceCRN, m.BaseDomain, hostname, destHostname)
+		if lastErr == nil {
+			return true, nil
+		}
+			return false, nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:308]: Nit: something wrong with the indent here
```suggestion
		return false, nil
```

File: pkg/asset/installconfig/powervs/metadata.go [hunk 5]
```diff
@@ -282,3 +285,84 @@ func (m *Metadata) GetDNSServerIP(ctx context.Context, vpcName string) (string,
...
+			Cap:      leftInContext(ctx),
+			Steps:    math.MaxInt32}
+
+	var lastErr error
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+		lastErr = m.client.AddSecurityGroupRule(ctx, vpcID, rule)
+		if lastErr == nil {
+			return true, nil
+			}
+			return false, nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:347]:
```suggestion
		}
		return false, nil
```

File: pkg/asset/installconfig/powervs/metadata.go [hunk 6]
```diff
@@ -282,3 +285,84 @@ func (m *Metadata) GetDNSServerIP(ctx context.Context, vpcName string) (string,
...
+
+func leftInContext(ctx context.Context) time.Duration {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return math.MaxInt64
+	}
+
+	duration := time.Until(deadline)
+
+	return duration
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:367]: Nit: 
```suggestion
	return time.Until(deadline)
```

**Review by r4f4 — COMMENTED** [review]
Nice refactor. Just a few minor comments.

/approve

**Comment by mjturek** [comment]
fixed!

**Comment by r4f4** [comment]
/hold cancel
#8331 has merged.

**Pull request merged**